Refactor `Table` and `Tree` to use `Column` objects internally

corranwebster opened the pull request:
This implements the first phase discussed in #4041 - a refactor of `Table` and `Tree` that pulls common logic into a `Column` object. This is a purely implementation detail at present, the `Column` objects are not exposed to the users in any way.

In particular, all current tests should pass unmodified.

Internally the Tree and Table have grown three new attributes:
- `_columns`: the list of column objects
- `_show_headers`: a bool that is true when no headings were passed
- `_data_accessors`: these are the accessors used when creating an ad-hoc `ListSource` from lists or dicts. The behaviour of accessors on column changes is a bit unclear, but looking at the code you can have more accessors than just those used in the headers. In future it might be good to make the behaviour a bit more specific, particularly around adding/removing columns (eg. should a completely new accessor be added to the list, where should it be added, and what does that do to existing data sources? Similarly for removal.).

This is at a state where it can be reviewed as a purely internal refactor. We probably want to expose some of these changes as part of the public API, but given that there is some complexity in this PR it's probably better to make those changes in a separate PR.

Ref #4041.
Fixes #4071.

To do:
- [x] Refactor `Table` and `Tree` interfaces.
- [x] Add `text`, `icon`, etc. methods
- [x] Tests for `Column` objects.
- [x] Make backend implementations use column interface.
  - [x] ~Dummy~ Nothing to do.
  - [x] Android
  - [x] Cocoa
  - [x] Gtk
  - [x] ~iOS~ Nothing to do, no Table or Tree widget.
  - [x] Winforms
- [x] ~Remove all mention of accessors from implementations (they're still used as column ids in a few backends, I think)~ Add an `id` property and use that where a column needs a name or identifier.

## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on core/src/toga/widgets/table.py:
```diff
@@ -75,28 +77,29 @@ def __init__(
-            self._accessors = list(accessors)
-        else:
+        self._missing_value = missing_value or ""
+        self._show_headings = headings is not None
```
This feels to me like it should be a property, rather than a writable attribute. Since it's a derived property, we need to make sure that it can't (however inadvertently) get out of sync with the intended value.

corranwebster replied:
I think it will end up being exposed as a read-only property - right now it's an internal flag which is set once during initialisation and is the source of truth about whether headings will be displayed or not. This matches the previous behaviour where if you set headings to None you could never add headings or change the table to show the headings.

freakboy3742 replied:
I'm not sure if there's value in making it public; but we can make that decision down the line.

freakboy3742 commented on core/src/toga/sources/columns.py:
```diff
@@ -0,0 +1,183 @@
+            case _:
+                return getattr(value, "icon", None)
+
+    def widget(self, row: Row[Value]) -> Widget | None:
```
Was omitting `widget` from the interface deliberate, or an oversight?

freakboy3742 commented on core/src/toga/sources/columns.py:
```diff
@@ -0,0 +1,183 @@
+        """
+
+
+class AccessorColumn(Column[Value]):
```
What other "Column" types are you anticipating here? I wonder if, from a naming perspective, it might make more sense to have `ColumnT` and `Column`.

corranwebster replied:
If the `Column` objects are exposed to the end-users, the immediate use-case would be for user-defined `Column` subclasses that work with the underlying row data more closely. In particular where the row objects are not necessarily objects with attrs, such as a DataFrame or a NumPy array and you don't want to copy the data into a `ListSource`.

There's also be a use-case for `Column` objects where the text and icon come from different attributes.

But I think most importantly, calling this `AccessorColumn` doesn't lock the API into making that the default "Column" class, given that part of the design intent is to open up the space to allow more complex things (like tool-tips or styling) where the current interface of str/tuple[icon, text]/widget is going to be difficult to extend elegantly. My *suspicion* is that the default column in the future is probably going to have separate accessors for text, icons, tooltips, etc.

I'm happy to call the `Column` protocol `ColumnT` if that's more consistent with other parts of the API.

freakboy3742 commented on core/src/toga/widgets/table.py:
```diff
@@ -268,30 +272,32 @@ def remove_column(self, column: int | str) -> None:
-        if self._headings is not None:
-            del self._headings[index]
-        del self._accessors[index]
+        del self._columns[index]
```
is there no need to clean up `_data_accessors` here?

corranwebster replied:
This is intentional - the link between headings and accessors is no longer via two synchronised lists, but via the columns.

But I'm not sure I properly understand what the expected behaviour of accessors were previously (that's what I'm getting at in the comment at the top of the PR). For ListSources of mappings everything is fine, because the Rows get their data from the keys of the mappings and as long as the column accessors match the keys, everything's good.

But for lists of sequences the mapping from accessors to rows is purely positional so you get situations like this:
``` python
data = [
    ((yak, "The Secret Life of Bees"), 2008, (green, 7.3), "Drama"),
    ((None, "Bee Movie"), 2007, (red, 6.1), "Animation, Adventure"),
    ...
]
# We don't show the last column initially
table = Table(headings=['Title', 'Year', 'Rating'], data=data) 
# table.data is now a ListSource where the row objects have 3 attributes: `title`, `year` and `rating`.
table.append_column('Genre')
# We've added a 4th column, but it's empty even though we gave the table 4 columns,
# because the ListSource was constructed with only three values.

# You can get around this by passing in 4 accessors directly:
table = Table(headings=['Title', 'Year', 'Rating'], accessors=['title', 'year', 'rating', 'genre'], data=data) 
# because then table.data has rows with all 4 accessors.
# But when we do:
table.append_column('Genre')
# it shouldn't add an additional 'genre' accessor at the end.
```
You can get something similar in the example app by doing: delete genre column, reset data, restore genre column:

<img width="752" height="592" alt="Image" src="https://github.com/user-attachments/assets/b0171b98-8916-45d3-96c0-0ce306aff4df" />

Or when you remove a column, things seem to work but then get odd when if you change the data:
``` python
table = Table(headings=['Title', 'Year', 'Rating', 'Genre'], data=data) 
# delete the rating column
table.delete_column('Year')
# all looks good, but `_accessors` is now `["title", "rating", "genre"]` so if you reload the data
table.data = data
# table.data is now a new ListSource but with only the three accessors and they no longer match the order. So you get years in the Rating column and ratings in the Genre column
```
With a tweak to the table example to delete years instead of Genre you can see this:
<img width="752" height="592" alt="Image" src="https://github.com/user-attachments/assets/f286c680-6848-49ae-87d9-fb33812b058f" />
(Change it to delete 'year', then press the delete button, then the 'reset data' button)

Neither of these behaviours seem right, but I'm not sure if that's a bug in the example and my expectation of how things should work For example, it might be reasonable to say that if you set the data using a list of sequences after changing the heading structure you need to make sure the order of the sequences matches the order of the columns.

So the _data_accessors is my attempt to square this circle: the idea is that it is the initial set of accessors passed in which tells the table the order of all known columns when given sequential data, independent of how the columns might change during the life-cycle of the app.

Again, this only affects data which is passed in as lists of sequences.

So I guess the question is what is the design intent?

freakboy3742 replied:
> So I guess the question is what is the design intent?

Fundamentally, the use case was to allow the Row/Node objects to store data that isn't necessarily rendered to the user in a column - e.g., an ID number for objects that are being represented in the table.  To that end, the interpretation you've taken with `_data_accessors` sees fairly close to the design intent - at time of construction, capture the attributes that will be retained in the Source. 

As you note, the gap you've found with adding/removing columns is almost entirely because the original data was specified as tuples/lists. Accessors have two purposes - identifying the name of the attribute to retrieve from the row; and determining how to map a tuple into a list of attributes. The second use case only exists when data is provided as a tuple or list; when a datum is provided as a dictionary, the dictionary keys give you the full list of attributes to preserve. 

I think I'm happy calling this an edge case that we can document - that the mapping of lists to Row is baked at time of construction, and if you want fully dynamic column addition/removal, you probably want to use a dictionary or custom Row. 

The only other thought I've got would be to replace `_data_accessors` with the accessor list *on the Source*. 

We make `accessors` an optional property on the ListSource that returns the list of accessors that are "known" to the Source. We create an empty ListSource when the Table is constructed using the accessors that come from construction. 

When new data is assigned, the list of accessors is preserved from the existing ListSource. If the user provides a custom Source, that logic is ignored; and if the user tries to replace a custom Source with an ad-hoc one, we raise an error (or maybe we build an accessor list based on the current columns). 

If the user tries to insert a column, we can validate if that column is on the list of known columns, and raise an error if it isn't. If the source doesn't provide an accessor list, then we don't do any validation.

That removes the need for the duplicated accessor tracking, as well as providing a way to give additional validation for the edge cases where a problem could exist.

freakboy3742 reviewed (changes requested):
The broad strokes of this look really good. I've flagged a couple of things inline. The biggest issue I can see is the `data_accessors` inconsistency; otherwise, the issues are mostly cosmetic/borderline bike shed, or likely oversights.

corranwebster commented:
I think everything is addressed:
- I've removed `accessor` from the `Column` protocol and replaced it with `id`, which for `AccessorColumn` is just the `accessor` value. The backend implementations now use `column.id` instead of `column.accessor` which means the way is clear for column objects that have different strategies for getting values from rows.
- I've changed `_data_accessors` to `_data_accessor_order`
- I've opened #4071 and #4072 to track the issues raised in the discussion of `_data_accessor`.
- I've added a note to the `Table` and `Tree` widget documentation explaining the behaviour of the accessors when columns are changed, and added a change log note that this PR resolves #4071.

freakboy3742 commented:
> I implemented it this way simply because the behaviour would be exactly the same as before this PR, for better or worse.

I see where that motivation comes from, but the internal identification of native widgets is very much an implementation detail, not a public API surface. I'm find if it changes between versions.

freakboy3742 reviewed (approved):
This is awesome. Even if we don't do any further work on Table/Tree columns, this is a significant cleanup and simplification of a bunch of duplicated logic; but it also provides a really solid launchpad for a better public API for Column representations, and it's pointed out a bunch of edge cases and bugs along the way. Thanks for all your work on this!
